### ISSUE TITLE
Reenable debugger

### ISSUE THREAD

**Issue by DirtyHairy** [stella-emu/stella#15]

**Comment by sa666666** [comment]
This is now partially done, but the TIA stuff is still broken.  I will update it as I go, but the code can at least compile and run with the debugger enabled.

**Comment by sa666666** [comment]
@DirtyHairy, I'm having an issue with implementing scanline tracing in the debugger.  The current code will advance one scanline each time the "Scan +1" button or Ctrl+l is pressed, but the issue is how it proceeds from one scanline to the next.

Currently, it tracks the current scanline, and stops advancing when it reaches the next line.  But what's happening is that when @ scanline 0, advancing immediately goes to scanline 34 (or whatever the first rendered line will be).  So it seems to skip all lines that aren't rendered.  That may be fine for rendering the image, but in the debugger needs to actually go through every line when advancing by frame, step or scanline.  The two former ones seem to be working, but the latter isn't.

Compare to the old core to see what I mean.  Also, for now, ignore the white square position, as it's incorrectly placed, and getting it working correctly depends on how we proceed with the aforementioned problem.

**Comment by DirtyHairy** [comment]
@sa666666 I'll look into it the next days. I guess we'll need to add some additional logic to FrameManager to do this properly.

**Comment by sa666666** [comment]
No problem, I'm working on other parts of the code ATM.

**Comment by DirtyHairy** [comment]
@sa666666 I just took a look; the problem is the semantics of `FrameManager::scanlines()`: at the moment, this will give the current frame scanline **if the current line is in the visible range of the frame** --- in all other cases, it returns the last total frame line count.

This is trivial to fix by changing the method to always return the current frame scanline, but `TIA::scanlines()` is used in several different places, so I'd like to ask you first of whether this should indeed be the correct behavior :smirk:

**Comment by sa666666** [comment]
The old semantics of scanline() was to return the actual scanline at the time of the method call, whether it was being drawn or not.  In other words, if I enter the debugger with a ROM of 262 scanlines, repeatedly stepping by scanline would go 0, 1, 2 ..., 262.  This assumption is used everywhere scanline() is currently used in the codebase, so nothing should break if you modify it to work this way (famous last words).

**Comment by DirtyHairy** [comment]
Commit 0efbbde fixes `TIA::scanline` to always return the current total scanline, as we discussed. This fixes stepping line by line. In addition, I have fixed `TIA::scanlinePos`, and the framebuffer preview seems to work properly now, including the "cursor".

@sa666666 , you can access the information you require by either
* `FrameManager::isRendering`: is true while drawing is in progress
* `FrameManager::getY`: gets the current visible line (aka scanlines() - frame offset)
As an alternative, we could also add callbacks to `FrameManager` (similarly to what I am doing for frame start and stop).

**Comment by sa666666** [comment]
Thanks @DirtyHairy, the scanline issues in the debugger are now fixed in https://github.com/stella-emu/stella/commit/c99cb33cda234940d5c04f1e8cc09a415405bac9.

**Comment by sa666666** [comment]
I just realized that rewind functionality in the debugger depends on state saving working, so this issue can't be closed until issue https://github.com/stella-emu/stella/issues/12 is done.

**Comment by DirtyHairy** [comment]
Hmm, I haven't touched anything related to state saving until now as I was expecting that the internal structure of the sprites would still be evolving and changing. However, it could well be that no more changes are needed and, in any case, stability of the serialized representation is not an issue for rewind, so I might take a look at what still needs to be done there in the future.

**Comment by sa666666** [comment]
The comment I left above was more for myself than for you.  I was simply stating that currently the state save/load is returning false, which doesn't allow rewind to work (it only functions when the current state is accurately saved and returns true).  I am waiting until you finish the implementation details before I get to it; you don't have to do anything with the state save stuff yourself.

**Comment by sa666666** [comment]
@DirtyHairy, is it possible to modify TIA::ystart() to return the actual ystart value being calculated, and not just the 0 when autodetect is used?  Or maybe introduce another method??

Basically, I need a method that returns the actual scanline number based on the TIA image shown in the debugger.  For example, the first line shown in the debugger is probably 34, 35, etc, depending on what the frame offset it.  For the first line (line 0 shown), I need a method that tells me that it is (for example) scanline 34.  I'm guessing this is calculated somewhere inside FrameManager, but is it actually ystart that I'm looking for?  And would modifying the ystart to be that value when it was initially 0 throw off something else in your algorithm?

**Comment by sa666666** [comment]
I've just experimented, and it's actually VblankManager::myLastVblankLines that I need.  FrameManager::scanlines() won't work, since before we do a 'fill to scanline' command, that method will return the actual # of lines, not the first line drawn.

I will experiment further to make sure that changing ystart() method to return myLastVblankLines won't cause any regressions.

**Comment by DirtyHairy** [comment]
I see --- I misunderstood you there :smirk: I am pretty sure that my code doesn't rely on ystart() anywhere, so changing its semantics should be fine from my point of view.

**Comment by sa666666** [comment]
Selecting the specific scanline in the TIA display with the mouse now works correctly, fixed in https://github.com/stella-emu/stella/commit/b5b058c615c4cf8a39dc1ed03bc4f7ccc4634b01.

**Comment by thrust26** [comment]
Were are we here? 

IMO, as soon as the debugger is working again (including save states), a final Stella 5.0 can be released. Or do I miss something? (PAL color loss would be nice)

**Comment by sa666666** [comment]
There are still a few things left TODO for the debugger.  Save states are easy; it's just that the TIA classes need to be finalized before I do it.  But there are still some issues I'd like to see closed before a final 5.0 release, including https://github.com/stella-emu/stella/issues/63, https://github.com/stella-emu/stella/issues/83, https://github.com/stella-emu/stella/issues/93 and https://github.com/stella-emu/stella/issues/108.

**Comment by thrust26** [comment]
Are the TODOs for the debugger listed somewhere? When would it make sense to start testing the debugger?

**Comment by sa666666** [comment]
Only in the code :smile: In particular, there's a bug of FIXME's in src/debugger/TIADebug.cxx.  To fix them, I need answers to questions posed in https://github.com/stella-emu/stella/issues/73.

**Comment by sa666666** [comment]
I'd also like to have the BUS and CDF bankswitch schemes in before the 5.0 release.

**Comment by DirtyHairy** [comment]
> The second one in particular is a regression from Stella 4.7.3. In general, I have no problem with releasing 5.0 with the new TIA not quite complete, but I'd like to not have any regressions.

My five cents: I do not think that TIA emulation will ever be complete :smirk: There will always be edge cases in which the emulation will differ from real hardware. A gate level simulation would be required to get even close to that, and even then, there would be effects (e.g. due to leak currents) that are not part of the emulation (just look at the documented cases where temperature changes what the TIA displays). I think that the new core is already much closer to hardware than the old core was, and there is still alot potential for getting even closer, but it will never achieve a 100% match :smile:

Philosophy aside, I agree about #83: I would very much like to get this closed before a release of 5.0. I am less enthusiastic about the other issues, though :smirk: #63 and #93 do not have any defined scope. I opened them to track the efforts of getting closer to real hardware in these cases, but there is no point at which you could add the tag "complete", there's always going to be the other odd case that we didn't examine. As for #108, I am very curious about the solution of this puzzle, but I don't think it is a show-stopper for a release, I guess the difference is mostly academic.

**Comment by sa666666** [comment]
I agree.  I'd like to see https://github.com/stella-emu/stella/issues/83 done, and BUS/CDF bankswitching added.  And the debugger stuff needs to be completed.  IMO, those are the real showstoppers for 5.0.  But depending on how hard https://github.com/stella-emu/stella/issues/83 is to fix, I'm not opposed to releasing 5.0 with it not fixed.

Overall, I'm pretty flexible with all this.  My main concern right now is finishing the debugger.  So if you both wouldn't mind looking at the FIXME's mentioned above and advising on the semantics of some of the functions ...

**Comment by DirtyHairy** [comment]
#83 is hopefully history :smirk: I would opt for ignoring #63 and #93: unless we have any issues connected to them, they are purely academic.

@sa666666 I'll try too look at those FIXMEs this weekend. As a sidenote, I think I have spotted a bug in the debugger: the "cursor" starts at hclock 0, not at 68 or 76. To check this, just look for a WSYNC; the cursor will immediately start moving right as you step over it.

**Comment by sa666666** [comment]
I've removed a few of the FIXME's in https://github.com/stella-emu/stella/commit/90aa152f596189c0b4e1eed362434fdaec9ee882 and https://github.com/stella-emu/stella/commit/4b56a178d8e7589da3a46256f04a3b08f448c759.  The remaining ones deal specifically with cases where one can't just change a register in the TIA with a POKE, but need to access the TIA sub classes directly.  That's the ones I need help with.

**Comment by DirtyHairy** [comment]
Unfortunately, I am rather low on time to spare at the moment, but I have finally managed to take a look at `TIADebug.cxx` and at the TODO. 

Concerning querying / setting the sprite positions: this information is not directly available, but instead encoded in the phase between the sprite counters and the playfield counter (myHctr). I can easily provide getters and the setters on the sprites for this. Concerning semantics: setting the position would set the sprite counter to its proper value and be esentially similar to strobing RESx on a suitable clock, with the same effects on draw counter decoding. In case of the players, I propose define the position as referring to NUSIZ 0 --- this will lead to a one pixel shift for double and quadruple width players, but this just reflects the fact that there is no well defined player position for players that is independent of NUSIZ :smirk: @sa666666 , @thrust26, what do you think?

Concerning register peeks / pokes (including those that are disabled atm): reading back the register values the way it is implemented now is problematic and may prove misleading as there usually is a clocking delay between the poke on the bus and the actual effect. I.e., the current implementation will not reflect effect of a `STA GRP0` after the instruction has finished, but only another pixel clock (which amounts to stepping another instruction). I would propose to implement a 128 byte "register file" that records all pokes and that can be read at any time to reflect the value of the last write. As a plus, we could drop all register-specific getters from the TIA and sprite classes and instead have a single getter paramterized by the register enum. Again, @sa666666 , @thrust26 , what do you think?

**Comment by sa666666** [comment]
I have no issues with anything you said.  In particular, having all the registers (whether read-only or not) in a separate array is similar to how 4.x worked, except in that case they were all in separate variables.

I defer to Thomas at this point, since he is a 2600 developer that would likely be using the debugger during development, so whatever makes sense from that POV should be what we do.

**Comment by thrust26** [comment]
I think I would prefer if the sprite position value represents the actual position, especially because it represents not a register, but the result of positioning the sprites.

Regarding the clocking delay, how about making this transparent? E.g. by displaying the current delay value and/or by making a poke a two step transition. So when value of a register is changed, the new pattern and the old pattern are displayed (e.g. by different shading) in parallel.

Also, how about allowing a pixel clock step additional to a CPU clock step?

**Comment by DirtyHairy** [comment]
> One question, though. Wouldn't we want to know the most current value of each register at any point in time? So would this 'register file' reflect that? I'm assuming it would, since you'd be using it to replace all getters in the TIA class.

I'm not sure that I understand your point correctly, but I think the "register file" would do exactly that: track the last write to each register.

> Would you mind doing an outline of this, or perhaps implementing it entirely? When you have time, of course.

I think I can implement it this week. I've got long train rides coming up on wednesday and thursday, and I should have ample time for coding there :smirk:

> I think I would prefer if the sprite position value represents the actual position, especially because it represents not a register, but the result of positioning the sprites.

Then I'll implement it this way :smile: For the players, this will imply that the position shifts when player width is changed between normal and wide.

> Regarding the clocking delay, how about making this transparent?

That's an interesting idea. The necessary information could be extracted by examining the delay queue. I think that this would make a nice feature in one of the next releases. @sa666666 , what do you think?

> Also, how about allowing a pixel clock step additional to a CPU clock step?

Unfortunately, I don't see any way to do this in Stella (either pixel clocks or CPU / RIOT clocks). In 6502.ts, the emulation is driven in a loop cycle by cycle, with each component receiving clocks in turn. This makes stepping very easy as it just means single stepping the loop, but it makes cycle exact bus access in CPU emulation a nightmare (which is why this currently not implemented properly). In Stella, the emulation is driven by the CPU whose smallest unit of execution is a single instruction. Each bus access increments the system clock by one, and if hardware registers are accessed, the peripherials fast-forward their internal state to reflect the current system clock. This is more efficient (and you get cycle exact bus timings for free), but the smallest unit by which the emulation can advance is a single instruction.

**Comment by sa666666** [comment]
WRT the part you asked me about, I would just need a mockup of what you'd like the UI to look like.  Again, from the POV of a 2600 developer, who would find it most useful.

As for the current code, I think I will do a test6 release in the next day or so, since we now have BUS/CDF bankswitching and state saving working.  Any objections to this?

**Comment by DirtyHairy** [comment]
I'm fine with a new prerelease :smirk: As for the UI, I'd like to defer this to @thrust26 as he is much more of an 2600 developer than me.

**Comment by thrust26** [comment]
Are we talking about the clock delays here? 

For the delay value, I would prefer it to be positioned directly before or behind the respective register (e.g. the GRP0 pattern or the M0 enable checkbox). 

For the two step transition:
- The GRPx and PFx pattern could get two intermediate pattern; one for getting enabled and one for getting disabled, e.g. a small rectangle in the middle when getting enabled and the inversed graphics when getting disabled. 
- The checkboxes could get intermediate colors; again one for getting enabled and one for getting disabled or the same graphics as for the pattern.

BTW: Regarding the current UI:
1. IMO the GR: prefix is quite superfluous.
2. We should try to get rid of the colons. The layout (spacing, alignment etc.) should make it obvious where a label belongs to.
3. The TIA and Audio tab could be merged (I prefer as much information at once as possible).
4. A small vertical gap between BL and PF would structure things a bit nicer 
5. The PF and GRPx pattern are handling the background differently. IMO it should be transparent for both. Else you have to check the colors first, before you know which PF registers are enabled.
6. The bits of the PF registers should be annotated, because I always have to look up the info :)  (e.g. 4..7, 7..0, 0...7). Either inside the boxes or above as labels.

**Comment by sa666666** [comment]
If possible, please draw how you'd like it to look and provide a picture.  Doesn't have to be neat, but I have problems picturing stuff from text descriptions only.

Perhaps draw something on paper and take a picture of it, do a mockup in a graphics editing program, etc.

**Comment by DirtyHairy** [comment]
I'm not sure that this is the right tool for this purpose, but the 30 days trial is free, and it usually does the job when we need it at work: https://balsamiq.com/ :smirk:

**Comment by thrust26** [comment]
I took the easy way and partially modified the TIA tab (mainly P0, M0 and PF).
![ui](https://cloud.githubusercontent.com/assets/24360774/24379093/cc4371ec-1345-11e7-975f-337278a34e3e.png)

**Comment by ale-79** [comment]
Would it be possible to show the content of both the normal and delayed registers for each player (and the delayed "enable" bit for the ball) and have a visual indication of which one will be displayed?

I think the vertical delay is one of the aspect of the TIA that's particulary difficult to understand while learning 2600 programming and a visual representation would greatly help.

**Comment by thrust26** [comment]
That's a good idea. We probably should put the delayed registers directly below the normal ones. For GRPx this is straightforward, but I think it would be a waste doing the same for BL.

**Comment by DirtyHairy** [comment]
I must admit that I have a hard time imagining out how to squeeze the current value, the pending value and the delay for all registers (into the widget). @thrust26 , I like your suggestions for checkboxes and bit masks; perhaps we can do something similar for numeric quantities by showing their upcoming value next to the current one in a different color (in the same field)?

As for the delay, perhaps we can skip it for now? The highest value is 6 color clocks (HMOVE), all others are one or two clocks, so it's only visible immediately after the corresponding instruction anyway.

As for visualizing the "old" and "new" GRPx registers: I think this is an excellent idea.

**Comment by thrust26** [comment]
I have no idea how we should display two values in the same field. 

As for the delay, since it is (usually) only valid for the current TIA write, how about displaying one, labeled value instead? So when e.g. a "sta GRP0" has been executed, this would be displayed as "GRP0: 1 px delay". And at the same spot on the UI, a change e.g. of PF2 would be displayed. And when there is no pending change, nothing (or an alternative text) is displayed.

**Comment by sa666666** [comment]
We do have some amount of vertical space, so extra widgets can be added.  I'd rather do that than try to cram everything into one widget, and make it very difficult to understand what it's actually implying.

Again, I will need pictures before I can implement any of this.  So I leave it to you to discuss until you come to an agreement ...

**Comment by DirtyHairy** [comment]
@thrust26 I was think of displaying values next to each other, but I think your second suggestion is much better. In fact, this matches what is going on below the hood: there is a delay queue which, for each of the next few cycles, holds the writes that are queued to execute in this cycle. We can just display the contents of the delay queue below the existing widgets, something like

     Queued write:
         * 0x47 -> GRP0, 1 pixel delay

As the maximum delay is 6 clocks, the maximum possible queue depth is limited to two entries (if HMOVE is hit with a RMW instruction), so two lines should suffice. For starters, we can just display the hex value of the write and the delay, but we can evolve this to showing a processed version of the induced change, e.g. the new pixel mask.

@sa666666 , if we agree on this, I think I can try to create an initial version of such a delay queue widget (as I know the delay queue code best :smirk:).

**Comment by sa666666** [comment]
If you wish to tackle the UI code yourself, then I won't object.  But it can get somewhat tedious, so I offer to do it if you find it too much work.

**Comment by thrust26** [comment]
Here is my latest mockup. Either we have individual delays or one combined one. I put in both, the combined one is at the bottom, maybe we find a better place.

Other changes: 
- darkened read only fields
- moved the labels closer to the checkbox
- removed all colons
- shortened, removed and fixed some labels

![ui3](https://cloud.githubusercontent.com/assets/24360774/24455923/9d7461a8-1491-11e7-85a8-2611cda87aac.png)
I like the "Queued write" prefix. Not sure if have to annotate the new value.

**Comment by DirtyHairy** [comment]
@thrust26 I think we should add the new value to the queue widget. It's redundant for the bitfields if we annotate them this way, but we need the value for the other registers which are not visualized this ways. Also, I must admit that I am not fully convincedyet  that this way of annotating the bitfields will not prove to be more confusing than helpful :smirk: If this does not work out well, then we could always add the upcoming bitfield to the queue widget.

@sa666666 I'll try my hands on the queue widget and see how it goes. If I get lost, I'll come running to you :smirk:

**Comment by thrust26** [comment]
Let's see what you can deliver :)

**Comment by sa666666** [comment]
The major issue you may run into is that the UI very much exists only to service current needs.  IOW, it's not a general purpose UI toolkit, and you will probably have to create some items from scratch, modify others, etc.  So it's not as easy as looking up the item to use and adding it; you may need to design it first :smile:

**Comment by DirtyHairy** [comment]
With commit 988b360, all FIXMEs from TiaDebug should be addressed --- shadow registers and positioning are implemented and work. There still is something funky with double buffering, but that'll have to wait until my trip home tomorrow :smiley:

**Comment by sa666666** [comment]
Read-only fields are now coloured differently (currently using the above colour, but this can be changed) in https://github.com/stella-emu/stella/commit/3e5df9c7bc4485f5472f409e0909004051d655ea.

**Comment by thrust26** [comment]
Probably a color in the middle between the bright input field background and the tab background would be perfect.

**Comment by sa666666** [comment]
Feel free to experiment and send me the colour you want :smile:   When I said "above colour", I meant the grey-ish one you used in the mockup.  It is very easy to change, just let me know what you want (the hex triple for RGB would be fine).

**Comment by thrust26** [comment]
220 207 165 should be perfect.

**Comment by DirtyHairy** [comment]
To chime in: the following screenshot is the delay queue widget that I've come up with. The widget is done and commited, but the content still is static mock text. I've got a bunch of changes lying around that'll tie the widget to the actual queue once they're finished, but I'd like to hear your opinion on whether I am going in the right direction with this first :smirk: @thrust26 , @sa666666 , what do you think?

![stella_delay_queue](https://cloud.githubusercontent.com/assets/2444224/24582368/f31891ac-172e-11e7-89d7-b4e0bc8ab8da.png)

**Comment by thrust26** [comment]
Looks OK to me.  Not sure about the order and labeling, but I have no better idea. :)

IMO it should be above the "unused TIA pins" option. And in read-only color of course.

**Comment by DirtyHairy** [comment]
Yeah, you're right :smirk: I originally put this below "unusued TIA pins" as didn't want to take up more visible space than absolutely necessary, but my addition of a background renders this moot, I guess :smile: I'll change the placement once I finelize it (and change the color to read-only).

**Comment by DirtyHairy** [comment]
The delay queue widget is finished in 3df4496 . That what it looks like:

![delay_queue_widget](https://cloud.githubusercontent.com/assets/2444224/24676069/d67dbe00-1981-11e7-95f3-66636fece315.png)

**Comment by thrust26** [comment]
It's a bit wide, isn't it?

**Comment by thrust26** [comment]
@sa666666: I still would prefer that the PF data is displayed over a transparent background.

**Comment by sa666666** [comment]
Here's with the latest changes:
![screenshot_20170404_174114](https://cloud.githubusercontent.com/assets/16061066/24676604/374a5bd6-195e-11e7-8204-42d2aa4fc04d.png)

I agree that it doesn't need to be that wide, since the longest TIA register string is 6 chars, the address is set at 2 chars, and the clock is 1 char at most.  Also, this could then be moved up and to the side of the "Queued Writes" string, since although we have some vertical space, it shouldn't be wasted (in particular, there are now issues with using the smallest font/window size that may force us to increase the minimum resolution for the debugger).

Thomas, I'm not sure what you mean by "PF data is displayed over a transparent background".  In the above, COLUBK is set to black, so it only makes sense to have that same colour displayed in the PF data??

**Comment by thrust26** [comment]
Why should we display COLUBK here? This is the PF register. 

Also, without looking at the values of COLUBK/COLUPF I don't know which bits are set. And if both are set to the same color (e.g. 0, to hide the PF temporarily), I can't recognize the bits at all. And finally it is inconsistent, because there is no COLUBK displayed in GRP0/1.

**Comment by sa666666** [comment]
OK, that makes sense.  And easy enough to fix.

**Comment by sa666666** [comment]
How about this?
![screenshot_20170404_180538](https://cloud.githubusercontent.com/assets/16061066/24677422/5a0136ec-1961-11e7-81e5-d2da22a075f9.png)

**Comment by thrust26** [comment]
Yup.

How about the other changes I suggested and painted above? (no colons; some label renaming; delayed register display...)

**Comment by sa666666** [comment]
We hadn't decided on who was going to do it, myself of @DirtyHairy.  I'm more familiar with the UI code, so it probably makes sense for me to do it.

My only issue is that I don't know how to interpret what you have in the Px and PF registers above.  In particular, what does it mean for the square to be hollowed out, and then sometimes that smaller inner square is present?

![screenshot_20170404_181531](https://cloud.githubusercontent.com/assets/16061066/24677839/c14f6700-1962-11e7-899e-64e8f7606b82.png)
![screenshot_20170404_181546](https://cloud.githubusercontent.com/assets/16061066/24677840/c159be6c-1962-11e7-9774-7da7091b06e5.png)

**Comment by thrust26** [comment]
That was an idea to display a delayed register change. But probably that's too much and we only use the delay queue instead.

**Comment by sa666666** [comment]
OK, so the squares are always full (or not there at all).  But what about the 2 P0 and 2 P1 graphics.  I assume you still want this, to show the old and new GRPx contents?

**Comment by DirtyHairy** [comment]
I agree, this widget is slightly too wide, I will change this.

**Comment by DirtyHairy** [comment]
Regarding the shadow stuff: I think what we'd like to have is a display of both the "old" and "new" player graphics registers. Writes to GRPx always end up in the new register, and the content of "new" is transferred to "old" once the *other* GRPx is written (so GRP1 triggers a transfer for player 0 and vice versa). The only difference between VDELx on and off is whether the contents of "new" (delay off) or "old" (delay on) are drawn. I would propose to add getters for the two different registers to `Player` and to display both on the widget, highlighting the register that is currently selected by VDELx. @sa666666 , I would prefer if you do the UI for this, but I can assist you with the getters and whetever else you may need.

**Comment by DirtyHairy** [comment]
As for visually displaying the "upcoming" GRPx: I have a vague idea of extending the DelayQueue widget to be smart enough to display additional information depending on the register written, i.e. the bitfield in case of GRPx of PFx. I think this would be easier to grasp than the two different types of boxes. @thrust26 , @sa666666 , what do you think?

**Comment by thrust26** [comment]
IMO that would mix two unrelated delays.

**Comment by DirtyHairy** [comment]
@thrust26 I am confused, this *is* what I was trying to say before :smirk: To rephrase:

I suggest to show visualize the effect of VDEL by displaying both "old" and "new" GRPx registers on the TIA tab, right there where GRPx is currently displayed. With "old" and "new" I am referring to the internal registers of the Players (as described in the programmer's guide), unrelated to the write delay queue.

What I am proposing to add to the queue widget is just a visualization of the upcoming value in the queue.

**Comment by thrust26** [comment]
OK then. I thought you meant the latter as an alternative.

**Comment by sa666666** [comment]
There are still some issues I'm not clear on, but for now I'm going to go ahead and implement the UI as you described.  We can tie in the functionality later.  Signing off for tonight.

**Comment by ale-79** [comment]
Here is a mockup I made showing both GRPx for each sprite and both ENABL:
The checkboxes before the labels indicate which one of the registers will be displayed and the state of the corresponding VDELxx bit. If VDELxx is set, the delayed register will be displayed, else the regular one.
Maybe the ball could be shown on a single line to save space.
![debug2](https://cloud.githubusercontent.com/assets/24704739/24701521/0e3b9d0c-19fb-11e7-9b53-68d1f229f3d6.png)
It would be nice if a change to a register could be visualized in some way, for example by outlining it.
This other mockup shows a write to GRP1, and the consequent copy of GRP0 and ENABL into their respective "delayed" registers.
![debug3](https://cloud.githubusercontent.com/assets/24704739/24701540/1bc53190-19fb-11e7-9719-b14a3718f7d7.png)

**Comment by DirtyHairy** [comment]
Thank you, @ale-79 , that's precisely what I had in mind. Some minor remarks:

* I'd add labels to the two different registers. "Old" and "New" is what is used in Andrew Tower's notes, the Programmer's manual just calls them 1 and 2. I'd opt for something like "Front" (aka new aka 1) and "Back" (aka old aka 2).
* If we both have the same thing in mind, than the red box is placed wrong for P1 imho.
* Of course, you're right about the ball, it's the same deal there. This is implemented, but I did never test this conciously. Does anyone know a ROM that actually uses this feature?

**Comment by ale-79** [comment]
In some revisions of the TIA programming guide there's a diagram showing the vertical delay logic and the registers are called "GRPx" and "GRPx Delay" there, while the "VDELPx" register is represented as a switch selecting which of the two register will be used as output.
![2600_tia_1a](https://cloud.githubusercontent.com/assets/24704739/24706155/0bfa1a96-1a0f-11e7-98f3-1eb9673aca96.png)
![jan_programming_guide](https://cloud.githubusercontent.com/assets/24704739/24706164/13552042-1a0f-11e7-9e67-a11943338bc5.png)



If I understand correctly how vertical delay works, "Px" in the mockup corresponds to the "first"/"new"/"regular" register, while "VDELPx" is the "second"/"old"/"delay" one. Writes from the processor always apply to the "first"/"new"/"regular" register, so when you write to GRP1 the first register is outlined. Because of that write, the value in the "first"/"new"/"regular" GRP0 register is copied into the "second"/"old"/"delay" GRP0 register, so the latter is outlined (the same happens for the ball). This happens regardless the status of "VDELP0", "VDELP1" and "VDELBL".

The checkboxes before the registers' labels indicates which one is selected as output and they only change when you set/unset the corresponding "VDELxx" bit. 
Since the "second"/"old" register is only selected when "VDELxx" is set, I thought it would make sense to directly label it as "VDELxx".

**Comment by sa666666** [comment]
OK, since the UI has changed (again) :smile:, I'll wait for another mockup that takes everything into account.  I have no issue with implementing any of this; what I don't want is to spend the time and commit the code, and then have to change it all again.

So I leave it to you guys to come up with a final UI that satisifies everyone, and then I'll implement it.

**Comment by thrust26** [comment]
Alex's is fine. I would only avoid repeating P0/P1 and BL (we can't/don't do this for Pos, HM, Reflect and Reset too). Also we should decide if we want to use register names or not. I would like to avoid to use them both. I am for non-register names where possible.

So IMO the labels should be cut to VDel and NuSiz.

**Comment by DirtyHairy** [comment]
@ale-79 my bad, I overlooked the checkboxes in front of P0 / P1, and I mistook the frame for the marker that denotes the currently active register. Your understanding of VDELx mechanics is spot on (or, at the very least, matches mine and the implementation in the new core :smirk:). @sa666666 , I am fine with the mockup, too.

Which reminds me: there also is a delay between the write to GRPx  and the corresponding register swap for the other player (and possibly the ball). This is also handled by the delay queue, and I guess I should make this visible in the widget, too :smiley: Everybody fine with "Register swap [P0|P1|BL]"?

**Comment by sa666666** [comment]
OK, but I would still like someone (or some-one's) to provide a mockup of the final decision(s) we've made.  So Thomas's removal of colons, name changes, etc.  I won't be starting the UI work until I know we have consensus on this, and have something to go by.  Until then, I'm working on other parts of the debugger (prompt, etc).

**Comment by ale-79** [comment]
A few other variations:

1: same as before but with labels renamed as suggested by @thrust26
![debug_1](https://cloud.githubusercontent.com/assets/24704739/24724684/5d08fe6a-1a4c-11e7-800c-61105cbd380f.png)

2 and 3: Checkboxes indicating the active register without labels, with or without additional "VDel" bit (it's redundant, but it might be clearer with it...)
![debug_2](https://cloud.githubusercontent.com/assets/24704739/24724696/693ca3a8-1a4c-11e7-861c-32b5f432b6b0.png)
![debug_3](https://cloud.githubusercontent.com/assets/24704739/24724702/6f0cca4c-1a4c-11e7-92ed-1f145232b7a3.png)

4: Active register with normal colors and inactive one greyed out. "VDel" bit next to the "old"/"delayed" register.
![debug_4](https://cloud.githubusercontent.com/assets/24704739/24724707/74841e3a-1a4c-11e7-9b70-8f38c21db63c.png)

With a bit of adjustements to the layout and by removing the colons as suggested, there should be enough room to add labels ("new"/"old" or similar) next to the registers, if needed.
I also agree with Thomas about annotating the PF bits, albeit I didn't include that in the mockups


> Which reminds me: there is a one cycle delay between the write to GRPx and the corresponding register swap for the other player (and possibly the ball), and this is also handled by the delay queue. I guess I should make this visible in the widget 😃 Everybody fine with "Register swap [P0|P1|BL]"?

Looks good to me.

**Comment by sa666666** [comment]
I actually prefer the last one, since having checkboxes before each register somehow doesn't look right (and sort of implies that the object can be toggled, which it can't of course).  I don't necessarily like the greying out of the registers, though.  What about just using a red outline around the currently active register, just like in a few posts above.  My reasoning for this isn't just looks; it is less complicated to add code to draw a red outline than to grey out the contents.

I don't really want to use checkboxes unless there's something that can be toggled; it shouldn't be used to indicate something else is enabled/disabled (at least from my POV).

**Comment by thrust26** [comment]
I too prefer the version without checkboxes. The way the content is displayed should signal the difference.

Also IMO there should be a visual difference for the currently inactive GRPx display between VDelPx enabled and not. Without VDel enabled, the inactive register never gets changed or displayed. However as soon as you enable VDel, it gets displayed. So that's very different from the way it works when VDel is enabled. 

How about drawing a cross (X) over each bit while VDel is disabled? And the inactive register should show no player color in any case. Instead I would use a darker version of the tab background color for the active bits. We just have to make sure that it is distinctive enough from the normal NTSC and PAL palette. 

![vdel](https://cloud.githubusercontent.com/assets/24360774/24727895/015e932a-1a58-11e7-9b92-296a0870e8c5.png)

**Comment by ale-79** [comment]
The way I undertand it, is that the second register is still changed even when Vdel is disabled. The only difference is that it's not displayed. It should be easy to verify it with  a test rom.

I like the use of a darker color for the inactive register.

**Comment by DirtyHairy** [comment]
For me, the important things are

1) Both registers are always visible and
2) That the currently selected register is visually tagged.

All the last suggestions from Alex and Thomas cover these, so, my two cents: do whatever is easier to do for you :smirk:

**Comment by sa666666** [comment]
OK, this is more doable.  One final question; should the VDel checkbox be read-only?  Because if so, it currently shows the wrong background colour.

In any event, I'll go ahead with the last mockup, and add the ability to draw an 'X' inside each position in the widget.  Once that is done, we can worry about how to tie the backend into the UI.  And for that part I will need some help ...

**Comment by DirtyHairy** [comment]
> The way I understand it, is that the second register is still changed even when Vdel is disabled. The only difference is that it's not displayed. It should be easy to verify it with a test rom.

That's definitely how it works. In fact, I had this wrong in 6502.ts first, and was "rewarded" with several glitched games (among others I think the font on the Video Cube title screen showed artifacts at the top)

> OK, this is more doable. One final question; should the VDel checkbox be read-only? Because if so, it currently shows the wrong background colour.

I am all for keeping it mutable --- it represents the register value, after all.

Take your time, Steve, and drop me a note when you need assistance with the backend. I might be a bit slower next week in responding as I will have to start preparing a guest lecture :smirk: All the best!

**Comment by sa666666** [comment]
I'm going to finish up on the debugger prompt before coming back to this, since I found major issues that need to be addressed.  So it may be towards the end of the coming week before I get back to this UI stuff.

**Comment by sa666666** [comment]
Just wanted to let everyone know that I'm about to finish up on this.  The last two weeks have been full of RL problems, so I didn't get much done so far.

**Comment by sa666666** [comment]
OK, the first pass of this is done in https://github.com/stella-emu/stella/commit/2d58825b251358706072b7250cd8de230ec5ea64.  Note that I haven't changed the missile/ball stuff completely yet, since I want to get the backend tied in for the players first.  So I await advice on how to do that.  Also, I still need to work on the togglewidget to have it draw 'X' (and shade the display), but again, I sort of need to tie the backend into the UI widgets to see it working in action.

![screenshot_20170424_123903](https://cloud.githubusercontent.com/assets/16061066/25344072/8c374284-28eb-11e7-8c1a-d873d787c250.png)

**Comment by SpiceWare** [comment]
I really like the TIA revisions.  Looks like most of the colons have gone away, just 2 left, which helps to declutter the form.  Should I remove them when I work on the Cartridge Tabs for BUS and CDF?

**Comment by DirtyHairy** [comment]
f51c6c9 adds support for showing queued old / new shuffles in the delay queue widget. ea7792d adds getters for the old / new GRP and ENABL values. @sa666666 , you can use those to connect the backend to the widget.

**Comment by thrust26** [comment]
@SpiceWare: Yes, please.

**Comment by thrust26** [comment]
@sa666666: 
This is coming a long quite nicely. 👍 

Can you remove the space between Pos and #?

**Comment by sa666666** [comment]
OK, it's easy enough to tie the getters into the new toggle widgets, but what about setters?  There are already setters for GRPx_new (the current register writes do this).  But what does it mean to write to the old registers?  Should we even allow it?  And if so, there needs to be methods for this too.

Also, would it be possible for someone to create a test ROM that allows me single-step it and see exactly how these new registers should behave?  Basically a very simple ROM that strips away everything else, and only illustrates behaviour for the GRPx new/old, and perhaps also EnaBL new/old.

**Comment by sa666666** [comment]
Something seems to have changed with the queued writes behaviour.  It is now occasionally showing references to VBlank and COLUP0 when those instructions aren't in the disassembly at all.  Perhaps something in https://github.com/stella-emu/stella/commit/f51c6c983bb8ff3185312904233bb9080da268d9 has changed??

**Comment by DirtyHairy** [comment]
Ouch, my bad, typo fixed in bd74b7d . Concerning setters: I see no harm in exposing both old and new registers for setting. I'll add a bunch of setters later tonight.

As for a test ROM: an easy way to see the shuffle mechanism in action is e.g. the 48px display on the title screen of Dark Chambers (or anything else that displays 48px bitmaps). Just step into the region where the bitmap is drawn, and you can see the GRP values being shuffled around. However, I don't have a quick testcase for the ball. I can try to put one together the next few days, though.

**Comment by DirtyHairy** [comment]
24e449c adds the necessary setters for modifying the "old" player and ball registers.

**Comment by sa666666** [comment]
OK, these are now tied into the UI, in https://github.com/stella-emu/stella/commit/f8d7253ec3d1c7f5b9c1b8b7e6117bc6456de194.  Next I need to add ability to blank the togglewidget contents (with light brown, as above), and also draw 'X' when necessary.  That leads to my next question.  Under what conditions should the widget use 'X' or 'browning out' the colours?  For example, in the screenshot below, with P1 VDel enabled, should the top or bottom be browned out?  And similarly, should the top or bottom be 'X'ed out??

![screenshot_20170425_194734](https://cloud.githubusercontent.com/assets/16061066/25410232/0bf04e1c-29f0-11e7-9450-dcbc69fe2ea0.png)

**Comment by DirtyHairy** [comment]
This is looking pretty good to me :+1: As far as I understood our discussion above, the register tht is currently unused should be crossed out and display its content in brown instead of player color. For the players that's "old" if VDel is active, and "new" otherwise.

**Comment by sa666666** [comment]
But that's not exactly what's explained above.  In one case, the register is browned out, and in another case it's browned out **and** crossed out too.  Code-wise, I have no problems doing this, but I need to know the logic of what happens under what circumstances.

**Comment by thrust26** [comment]
Yup, see my post 20 days ago.

**Comment by thrust26** [comment]
Minor issues:
- there are two colons left (Strobes:, Queued Writes:)
- the header above the collision boxes are not aligned correctly (there are a few other minor alignment issues, but I suppose those can wait until we have finalized everything)
- BL: "Delay" should be renamed into "VDel" (for consistency)
- M0/M1: Maybe rename "Reset" into "ResMP" (or something more obvious like "Reset to P"),
- How about reordering the strobes so that the left column contains RESxy identical to the collision display order? The right column would then be WSync, RSync, HMove, HmClr and CxClr (though this might more sense closer to the collision display)

**Comment by ale-79** [comment]
I think only one visual indication for the inactive registers (either browned out or browned out and crossed) should be used in all cases.

With VDEL enabled, the bottom (old) register is active (its content is colored, meaning that it's the one that will be displayed), while the top register (new) is inactive.
In the same way, with VDEL disabled, the top register is active while the bottom register is inactive.
That's the only effect of VDEL. That would be the position of the "switch" in the diagram from the TIA programming guide, which connect one of the two registers (top and bottom rows in the TIA tab) to the video ouptut. Once set, that switch doesn't move unless VDEL is toggled again.

> While VDELP0 is enabled:
> 
>     the top row is without color (signalling, this row is not displayed now but later)
> 

I think there's no real difference between this and the case of  VDELP0 disabled
It's just the **content** of the top register that "could" be displayed later, but only if it will be copied into the bottom row, which is the active register, and that will only happen if a write to the other player occurs before the top row is changed.
In all cases, the content of the active (colored) register is the only one that will be sent to the video output, while the content of the inactive register won't be displayed.


> In both cases, when a write to GRP0 happens:
> 
>     the current content of the top row is copied into the bottom row
>     the top row is updated with the new value



I lost you there. When you write to GRP0 the only thing that happens is that the top row is updated with the new value.
The top row is copied into the bottom one only when a write to the **other** player (GRP**1** in this case) happens.

**Comment by thrust26** [comment]
I corrected the result of a write to GRP0. Thanks for pointing out my mistake.

If I get you right, we are discussing if the crosses make sense of not. IMO they do, because there is a difference. Either the disabled bits are displayed because they are getting copied into the display register or because VDEL is set from disabled to enabled. And the former is a repeating behavior e.g. inside a kernel, while the latter usually only happens once during setup code. Also the crossed out bits can (and usually will) be overwritten without ever being in displayed, while the non-crossed out bits will usually show up soon (with the next write to the other player). So in the latter case, the two player registers display a queue. With VDEL disabled, there is no queue and the disabled bits are (usually) just unused garbage.

That's my understanding, but I can live without the crossed out bits.

**Comment by sa666666** [comment]
I will go with the logic from @thrust26.  If we decide to remove the 'X' later, it will just be a removal of a method call.

As for 'HM' alignment, I haven't really started on the missile/ball yet, so there's no guarantee that the HM in those will be in the same position after I'm finished.

I'll remove the colons and align the collision labels soon.

As for renaming NuSiz to Size, for the players it represents both number and size, but for the missiles only the size (is that correct)?  If so, then the naming should be different.  But I need to know this now, since if we're aligning things, I don't want to have to change the text multiple times and realign everything again.

**Comment by thrust26** [comment]
You are right, NuSiz means more than just Size. So we should not touch those names.

**Comment by ale-79** [comment]
> If I get you right, we are discussing if the crosses make sense of not. IMO they do, because there is a difference. Either the disabled bits are displayed because they are getting copied into the display register or because VDEL is set from disabled to enabled. And the former is a repeating behavior e.g. inside a kernel, while the latter usually only happens once during setup code. Also the crossed out bits can (and usually will) be overwritten without ever being in displayed, while the non-crossed out bits will usually show up soon (with the next write to the other player). So in the latter case, the two player registers display a queue. With VDEL disabled, there is no queue and the disabled bits are (usually) just unused garbage.


I get you, but I think the difference is in how the registers are typically used in a game, and not in the way they actually work. Both can be either active or not, but if and when their content will be displayed (because they become active by setting VDEL, or because their content is copied into the other, active register) only depends on the code.

I don't know if it would make sense to do that, but suppose that in a kernel you have to display some copies of the players across a line, and they only need to use two different graphics. You could load the new and old register with the two graphics and them just switch VDEL to the proper value before each player copy. In that (improbable, I admit) case, having different ways to visualize the inactive register doesn't seem quite right.

Anyway, I absolutely agree that's just a detail, and I'm fine with either solution. I think the important thing is that you can see both registers and visually follow how they're loaded. Technically, the information about the active register is already conveyed by the VDEL checkbox. The browned out / crossed content is just an extra way to visualize it.

**Comment by sa666666** [comment]
Player 'old' registers are now activated in https://github.com/stella-emu/stella/commit/8c756001f88b0884460a8c8ff6a5ec6b05a70cda.  Also aligned the collision register labels, and moved over the strobe buttons, to make room for the 'fixed debug colours' area.  Snapshot shows both player 0 and 1 in various states, which I hope properly illustrates what we want.  Note that I went with horizontal lines instead of crosses, since there's actually no code in the UI toolkit to draw diagonal lines.
![screenshot_20170427_141124](https://cloud.githubusercontent.com/assets/16061066/25494493/66606eb2-2b54-11e7-8c49-d1d96c76b5aa.png)

**Comment by thrust26** [comment]
Looks good!

But why did you chose horizontal lines over X lines?

**Comment by sa666666** [comment]
Read the last sentence :smile:

**Comment by thrust26** [comment]
Oops!

**Comment by sa666666** [comment]
While I'm finishing this stuff up and moving buttons around, I'm wondering if the 5 ResXX buttons should be placed after the respective graphics objects?  Then CxClr will go into the collision area, and we only have 4 buttons left.  At that point, do we even need to label them as 'Strobes', and we can move them elsewhere on the UI??

**Comment by ale-79** [comment]
I think is a good idea to have them near the objects they're related to and also that there's no need to label them as "strobes" because they're already visually different.

**Comment by thrust26** [comment]
I agree. 

And if we can arrange the HM registers accordingly, the might be able to put HmClr nearby

**Comment by sa666666** [comment]
Current progress from https://github.com/stella-emu/stella/commit/9416d5212c8c60b9ddc1e0fbec9043f49b2b7ed7.  Fixed debug colours added, and buttons rearranged.  Please provide feedback, since I'd like to do another pre-release tomorrow and hopefully get to the main release soon.
![screenshot_20170502_185950](https://cloud.githubusercontent.com/assets/16061066/25640301/a81d51d0-2f69-11e7-930d-c1eb76317824.png)

**Comment by SpiceWare** [comment]
Looks good to me.

**Comment by ale-79** [comment]
I just compiled it. Looks great!

**Comment by thrust26** [comment]
Looks good, just a few **minor** comments:

- The CxClr button could be moved a few pixel down, to be better aligned vertically with the BL checkbox.
- I would like to have the ResXY buttons to be vertically aligned, where the ResPx buttons are now.
- the labeling before the debug colors should be shortened to match the other labels (P0, P1, M0, M1, BL, PF, BK). That way we can clean the UI a bit more. 
- also IMO the debug colors should be next to the normal colors
- and then we have three different register orderings in that area (colors, debug colors and collisions). If we keep the color registers order, then debug colors should be order like this:
  P0 M0
  P1 M1 
  PF BL
  BK HMove (spelling changed for consistency)

- Unless someone finds a better solution, the vertical order of the collisions can stay as they are. 
- The rows of debug colors and collisions should be vertically aligned to the color display rows.
 
BTW: **Maybe** it would make sense to put the debug color labels inside the color boxes?

But that can wait until the next release.

**Comment by sa666666** [comment]
What you're asking for isn't impossible, but it will really take some effort to do.  Particularly the alignment of buttons, since the current code lays stuff out left-to-right, top-to-bottom.  So we don't know where the buttons will appear until we get to ResP0, but at that point I've passed the collision registers.

The problem with re-arranging the debug colours is that they previously followed the 'ROYGBIV' scheme (see issue https://github.com/stella-emu/stella/issues/114).  So that will break the scheme if I change them.  What you're saying makes sense (they line up with the COLUBx registers), but what was said in https://github.com/stella-emu/stella/issues/114 also makes sense.  So I don't know how to proceed.

I will try to get most of this done, except for the 'maybe' items.  I want to get this completed and move on to other parts of the code, since you may have noticed that this issue already has over 110 comments!

**Comment by sa666666** [comment]
A compromise: leave collision registers and debug colour placement as-is.  Align the Rexxx buttons after each object (having CxClr not aligned doesn't matter IMO, since they're not related anyway).  Add other options as possible (checkbox for fixed colours is easy, perhaps crossing out unused colours is easy too).

**Comment by thrust26** [comment]
I knew about issue #114, and tried to keep the order still intact. Just read the debug colors line by line.

But let's hear what the others say.

BTW: Why did you chose to use mixed case names for the buttons (e.g. CxClr) initially? 

The constants for these registers are all completely uppercase (within source code and also Stella disassembler). I got long used to it, but it still looks a bit odd.

**Comment by sa666666** [comment]
I agree, lets hear more input.  I can align the ResXY buttons right now, and also change to all uppercase in the buttons (I don't recall the reason for doing it with mixed case).  But then I want to do a pre7 release.

**Comment by thrust26** [comment]
Do whatever doesn't delay the next pre-release. :)

**Comment by SpiceWare** [comment]
I really like Thomas' suggestion for moving/relabeling the debug colors.  A minor change would restore ROYGBIV:
![c9510f5e-2feb-11e7-9f5f-98641393dcac 2](https://cloud.githubusercontent.com/assets/26531922/25667486/3fa0d778-2fe9-11e7-9a12-19b2c47b0a14.png)

**Comment by thrust26** [comment]
Oops, looks like I swapped BL and PF. That was not intentional.

Or, alternatively we could keep my wrong order and just swap colors between BL and PF.

**Comment by SpiceWare** [comment]
The order of PF BL is logical as the ball color would be just below the missile colors.

**Comment by thrust26** [comment]
Agreed. Probably that was the reason for my mistake.

So, how about just swapping the two colors? Would that cause a problem?

**Comment by sa666666** [comment]
Since the colours have been redefined since Stella 4.x, I see no reason why we can't swap PF and BL ones.  And I too really like the layout of having them next to the COLUxx registers, since it reinforces the association between players and missiles, etc.  So the only difficult thing for me is aligning CxClr with the rest of the ResXY buttons.  But I'll see what I can do.  In any event, I want to do pre7 release this evening.

**Comment by thrust26** [comment]
That alignment with the other buttons is really only a 2nd thought, since I didn't know where to put the collision segment. So I decided it would look best aligned. 

So absolutely no big deal if it is not aligned.

**Comment by sa666666** [comment]
Latest changes in https://github.com/stella-emu/stella/commit/1d30f1ddc03feee7bfd814d635ea599f9f083d0d.  Every suggestion made is done, except for the second 'maybe' and the alignment of CXCLR.  I'll be doing pre7 in a moment.
![screenshot_20170503_195950](https://cloud.githubusercontent.com/assets/16061066/25684400/ccc615ee-303b-11e7-8c5d-9415f803bebe.png)

**Comment by thrust26** [comment]
One small wish for the next release (I missed this until testing now, sorry): 
The active bits in the inactive P0/P1/BL registers should be displayed darker (not brighter) than the background (see my small mockup 28 days ago). IMO that makes the disabled state more obvious.

And in the end, all (maybe with a very few exceptions) colons should be removed from all tabs and other parts of the GUI (e.g. Options menu).

**Comment by sa666666** [comment]
I simply used the background colour of non-editable widgets.  I guess it's easy enough to add another colour.

As for the colons, yes, I am in the process of removing them.  Perhaps you didn't notice a recent commit "the great colon purge"? :smile:

**Comment by sa666666** [comment]
Colour is changed in https://github.com/stella-emu/stella/commit/847e910d26360d9aba45ff208029a78635a8d23f to match one in post from 28 days ago.
![screenshot_20170504_094107](https://cloud.githubusercontent.com/assets/16061066/25702915/d653391c-30ad-11e7-9710-3df1c209f180.png)

If there are no other items, I want to close this issue soon.  The debugger has be re-enabled for quite some time, and what we've done here has greatly surpassed the original point of the issue.

**Comment by thrust26** [comment]
Yup, lets open a new topic for "beautification" or so :)

**Comment by sa666666** [comment]
Closing this since its goal was reached long ago.

**Comment by thrust26** [comment]
There is a small bug when enabling M0 and M1. Unlike for P0, P1 and BL, the real color is not displayed in the (former) checkbox immediately. The update only happens after the next instruction

BTW: Until now I missed that the checkbox was replaced with a color box, which works as a checkbox too. This is pretty unusual, and maybe irritating especially because there is no "Enable" label anymore. I would prefer to have at least the label back (I think the tab has enough horizontal space). And the behavior of the color box should be documented.

**Comment by sa666666** [comment]
Arrgh, not this issue again (it already has 138 replies)

As for the removal of the checkbox and label, that was at your insistence :imp: 
As is normal, I need a test case/specific way to trigger the bug, and a brief description of what should be happening.

**Comment by thrust26** [comment]
Sorry, I have learned that GUI refinements tend to be very tedious. Especially with multiple people with multiple ideas and preferences involved. But in our case, IMO the result was well worth it.
And no, it wasn't my "insistence". It was Alex idea and I didn't really notice the result until now. My bad, ignore me here. :) (and I think I will soon get used to it)

Anyway, back to the bug:
- start e..g River Raid and enter the debugger. Initially both missiles are disabled, displayed by the tabs background color. Fine.
- now double click* on the color box for M0. The color changes into a darker brown. This is not OK, I would expect the color to change into yellow (the color of P0) immediately.
- when you now step the CPU, THEN the expected color shows in color box for M0.

*BTW: why a **double** click? A single click would do. (please, don't kill me for this! :D)

**Comment by sa666666** [comment]
Insistence is perhaps a strong word.  In any event, I will look into it.

As for the single vs. double-clicking, I don't know if there's any technical reason for one or the other.  It should be easy enough to change it.

**Comment by ale-79** [comment]
Using color boxes for all the graphic objects allows to uniform the way the active register is visualized in objects with vertical delay.
The enable bit for missiles and ball sets/unsets the single pixel of the graphic object, and you can think the GRPx adn PFx registers as containing several enable bits, one for each of the pixels. So I think it makes sense using the same layout for all the objects. An "enable" label wouldn't provide additional information IMO, but I wouldn't mind if it's added.
As Thomas said, personal preferences play a large role here. 

I confirm the bug and the fact that it only happens for missiles, not players, ball or playfield.

**Comment by thrust26** [comment]
For me the color boxes are fine. Already got used to them. :)

**Comment by sa666666** [comment]
The last bugs mentioned by @thrust26 above are fixed in https://github.com/stella-emu/stella/commit/43c22af2742064a2974049c345e874b9c116ce03, so we can finally close this issue.
